File: 0268-missing-number/0268-missing-number.java
class Solution {
    public int missingNumber(int[] nums) {
        Arrays.sort(nums);
        int lo = -1; 
        int hi = nums.length;
        while (hi-lo > 1) {
            int m = lo + (hi-lo)/2;
            if (nums[m] == m) {
                lo = m;
            } else {
                hi = m;
            }
        }
        return hi;
    }
}

/*
class Solution {
    public int missingNumber(int[] nums) {
        HashSet<Integer> hs = new HashSet();
        for (int num : nums) {
            hs.add(num);
        }
        for (int i = 0; i < nums.length + 1; i++) {
            if (!hs.contains(i)) {
                return i;
            }
        }
        return -1;
    }
} */
// Time: O(n), Space: O(n)

File: 0268-missing-number/0268-missing-number.rs
use std::collections::HashSet;

impl Solution {
    pub fn missing_number(nums: Vec<i32>) -> i32 {
        let mut hs: HashSet<i32> = HashSet::new();
        let len = nums.len();
        for (i, n) in nums.iter().enumerate() {
            hs.insert(*n);
        }
        for i in 0..len+1 {
            let n = i as i32;
            if !hs.contains(&n) {
                return n
            }
        }
        return -1
    }
}